app/grandchallenge/evaluation/migrations/0077_create_evaluation_input_set.py

```diff
@@ -0,0 +1,117 @@
+# Generated by Django 4.2.20 on 2025-03-19 08:49
+import uuid
+
+import django.db.models.deletion
+from django.db import migrations, models
+
+from grandchallenge.evaluation.models import (
+    get_existing_evaluation_input_set_for_inputs,
+)
+
+
+def update_evaluation_inputs(apps, _schema_editor):
+    Evaluation = apps.get_model("evaluation", "Evaluation")  # noqa: N806
+    EvaluationInputSet = apps.get_model(  # noqa: N806
+        "evaluation", "EvaluationInputSet"
+    )
+
+    batch_size = 5000
+
+    evaluations = Evaluation.objects.prefetch_related("inputs").iterator(
+        chunk_size=batch_size
+    )
+    updated_evaluations = []
+
+    for evaluation in evaluations:
+        inputs = evaluation.inputs.all()
+
+        input_set = get_existing_evaluation_input_set_for_inputs(
+            model=EvaluationInputSet,
+            inputs=inputs,
+        )
+        if not input_set:
+            input_set = EvaluationInputSet.objects.create()
+            input_set.values.set(inputs)
+
+        evaluation.input_set = input_set
+        updated_evaluations.append(evaluation)
+
+        if len(updated_evaluations) >= batch_size:
+            Evaluation.objects.bulk_update(updated_evaluations, ["input_set"])
+            updated_evaluations = []
+
+    if updated_evaluations:
+        Evaluation.objects.bulk_update(evaluations, ["input_set"])
+
+
+class Migration(migrations.Migration):
+    dependencies = [
+        ("components", "0025_alter_componentinterface_overlay_segments"),
+        ("evaluation", "0076_alter_evaluation_unique_together"),
+    ]
+
+    operations = [
+        migrations.CreateModel(
+            name="EvaluationInputSet",
+            fields=[
+                (
+                    "id",
+                    models.UUIDField(
+                        default=uuid.uuid4,
+                        editable=False,
+                        primary_key=True,
+                        serialize=False,
+                    ),
+                ),
+                ("created", models.DateTimeField(auto_now_add=True)),
+                ("modified", models.DateTimeField(auto_now=True)),
+                (
+                    "values",
+                    models.ManyToManyField(
+                        related_name="eval_inputs",
+                        to="components.componentinterfacevalue",
+                    ),
+                ),
+            ],
+            options={
+                "abstract": False,
+            },
+        ),
+        # Add the temporary field with a different name
+        migrations.AddField(
+            model_name="evaluation",
+            name="input_set",  # Temporary field name
+            field=models.ForeignKey(
+                blank=True,
+                null=True,
+                on_delete=django.db.models.deletion.PROTECT,
+                to="evaluation.evaluationinputset",
+            ),
+        ),
+        migrations.RunPython(update_evaluation_inputs, elidable=True),
```

Comment:
This needs to be split up into 3 migrations as data migrations need to be separate from table migrations.

Reply:
Right, sorry, I had this as separate migrations initially, and then forgot why I did that and merged them into one. Sigh.